### PR TITLE
Adding servlet filter support for managing context

### PR DESCRIPTION
**Pull request by gr8routdoors**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -39,13 +39,20 @@
...
         <dependency>
             <groupId>org.immutables</groupId>
             <artifactId>value</artifactId>
             <version>2.1.9</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>3.1.0</version>
```
Comment (cchacin) [inline pom.xml]: `question`: optional dependency instead?

File: pom.xml [hunk 2]
```diff
@@ -100,6 +107,12 @@
             <version>20.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-test</artifactId>
```
Comment (cchacin) [inline pom.xml]: `nitpick`: I know this is a test dependency but wondering if we can achieve the same using just `mockito`

File: pom.xml (hunk 1 above)
Comment (brianm) [inline pom.xml]: Can you add a `<scope>provided</scope>`?

File: src/main/java/com/groupon/jtier/filter/CtxFilter.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                         final FilterChain chain) throws IOException, ServletException {
+
+        final Ctx requestCtx = defaultCtx.newRoot();
+        requestCtx.attachToThread();
+
+        try {
+            chain.doFilter(servletRequest, response);
+        }
+        finally {
+            requestCtx.cancel();
```
Comment (brianm) [inline src/main/java/com/groupon/jtier/filter/CtxFilter.java]: This needs handling for requests which switch to async. The Ctx should not be cancelled until the async request finishes.

File: src/test/java/com/groupon/jtier/filter/CtxFilterTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+
+public class CtxFilterTest {
+
+    MockHttpServletRequest req;
```
Comment (brianm) [inline src/test/java/com/groupon/jtier/filter/CtxFilterTest.java]: private

File: pom.xml (hunk 2 above)
Comment (brianm) [reply pom.xml]: Using spring-test here is fine IMHO, is much simpler than mockito :-)

Comment (cchacin) [reply pom.xml]: it's fine to keep this but I disagree on that for this particular case, here a lot shorter/clearer working test with `mockito`: 

```java
public class CtxFilterTest {

    private HttpServletRequest   req = Mockito.mock(HttpServletRequest.class);
    private HttpServletResponse  res = Mockito.mock(HttpServletResponse.class);
    AtomicReference<Ctx> ctx;
    private FilterChain chain;

    @Before
    public void setUp() {
        chain = (request, response) -> ctx.set(Ctx.fromThread().get());
        ctx = new AtomicReference<>();
    }

    @Test
    public void supportsSharedDefaultContext() throws Exception {
        // Configure
        final Ctx.Key<String> KEY = Ctx.key("foo", String.class);
        final String VALUE = "bar";
        final Ctx defaultCtx = Ctx.empty().with(KEY, VALUE);

        final CtxFilter filter = new CtxFilter(defaultCtx);

        // Run
        filter.doFilter(req, res, chain);

        // Verify
        assertThat(ctx.get()).isNotNull();
        assertThat(ctx.get().get(KEY).get()).isEqualTo(VALUE);
    }

    @Test
    public void supportsEmptyDefaultSharedContext() throws Exception {
        // Configure
        final CtxFilter filter = new CtxFilter();

        // Run
        filter.doFilter(req, res, chain);

        // Verify
        assertThat(ctx.get()).isNotNull();
    }

    @Test
    public void cleansUpTheRequestContextAfterTheRequest() throws Exception {
        // Configure
        final CtxFilter filter = new CtxFilter();

        // Run
        filter.doFilter(req, res, chain);

        // Verify
        assertThat(ctx.get().isCancelled()).isTrue();
    }

    @Test
    public void leavesTheDefaultContextIntactAfterTheRequest() throws Exception {
        // Configure
        final Ctx defaultCtx = Ctx.empty();

        final CtxFilter filter = new CtxFilter(defaultCtx);

        // Run
        filter.doFilter(req, res, chain);

        // Verify
        assertThat(defaultCtx).isNotNull();
    }
}
```

Comment (cchacin) [reply pom.xml]:
```diff
diff --git a/src/test/java/com/groupon/jtier/filter/CtxFilterTest.java b/src/test/java/com/groupon/jtier/filter/CtxFilterTest.java
index e572bfc..07c8c58 100644
--- a/src/test/java/com/groupon/jtier/filter/CtxFilterTest.java
+++ b/src/test/java/com/groupon/jtier/filter/CtxFilterTest.java
@@ -17,31 +17,26 @@ import com.groupon.jtier.Ctx;
 
 import org.junit.Before;
 import org.junit.Test;
-import org.springframework.mock.web.MockFilterChain;
-import org.springframework.mock.web.MockHttpServletRequest;
-import org.springframework.mock.web.MockHttpServletResponse;
+import org.mockito.Mockito;
 
-import java.io.IOException;
 import java.util.concurrent.atomic.AtomicReference;
 
-import javax.servlet.GenericServlet;
-import javax.servlet.ServletException;
-import javax.servlet.ServletRequest;
-import javax.servlet.ServletResponse;
+import javax.servlet.FilterChain;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
-
 public class CtxFilterTest {
 
-    MockHttpServletRequest req;
-    MockHttpServletResponse res;
+    private HttpServletRequest   req = Mockito.mock(HttpServletRequest.class);
+    private HttpServletResponse  res = Mockito.mock(HttpServletResponse.class);
     AtomicReference<Ctx> ctx;
+    private FilterChain chain;
 
     @Before
     public void setUp() {
-        req = new MockHttpServletRequest();
-        res = new MockHttpServletResponse();
+        chain = (request, response) -> ctx.set(Ctx.fromThread().get());
         ctx = new AtomicReference<>();
     }
 
@@ -54,14 +49,6 @@ public class CtxFilterTest {
 
         final CtxFilter filter = new CtxFilter(defaultCtx);
 
-        final MockFilterChain chain = new MockFilterChain(new GenericServlet() {
-            @Override
-            public void service(final ServletRequest req,
-                                final ServletResponse res) throws ServletException, IOException {
-                ctx.set(Ctx.fromThread().get());
-            }
-        });
-
         // Run
         filter.doFilter(req, res, chain);
 
@@ -75,14 +62,6 @@ public class CtxFilterTest {
         // Configure
         final CtxFilter filter = new CtxFilter();
 
-        final MockFilterChain chain = new MockFilterChain(new GenericServlet() {
-            @Override
-            public void service(final ServletRequest req,
-                                final ServletResponse res) throws ServletException, IOException {
-                ctx.set(Ctx.fromThread().get());
-            }
-        });
-
         // Run
         filter.doFilter(req, res, chain);
 
@@ -95,16 +74,6 @@ public class CtxFilterTest {
         // Configure
         final CtxFilter filter = new CtxFilter();
 
-        final MockFilterChain chain = new MockFilterChain(new GenericServlet() {
-            @Override
-            public void service(final ServletRequest req,
-                                final ServletResponse res) throws ServletException, IOException {
-                ctx.set(Ctx.fromThread().get());
-
-                assertThat(ctx.get().isCancelled()).isFalse();
-            }
-        });
-
         // Run
         filter.doFilter(req, res, chain);
 
@@ -119,14 +88,6 @@ public class CtxFilterTest {
 
         final CtxFilter filter = new CtxFilter(defaultCtx);
 
-        final MockFilterChain chain = new MockFilterChain(new GenericServlet() {
-            @Override
-            public void service(final ServletRequest req,
-                                final ServletResponse res) throws ServletException, IOException {
-                ctx.set(Ctx.fromThread().get());
-            }
-        });
-
         // Run
         filter.doFilter(req, res, chain);
```

**Comment by gr8routdoors** [comment]
All comments should now be addressed.

**Pull request merged**